core/src/apps/common/sdcard.py

```diff
@@ -45,17 +45,55 @@ class SdProtectCancelled(wire.ProcessError):
     return await confirm(ctx, text, confirm=btn_confirm, cancel=btn_cancel)
 
 
+async def format_card_dialog(ctx: wire.GenericContext) -> bool:
+    # Format card? yes/no
+    text = Text("SD card error", ui.ICON_WRONG, ui.RED)
+    text.bold("Unknown filesystem.")
+    text.br_half()
+    text.normal("Use a different card or")
+    text.normal("format the SD card to")
+    text.normal("the FAT32 filesystem.")
+    if not await confirm(ctx, text, confirm="Format", cancel="Cancel"):
+        return False
+
+    # Confirm formatting
+    text = Text("Format SD card", ui.ICON_WIPE, ui.RED)
+    text.normal("Do you really want to", "format the SD card?")
+    text.br_half()
+    text.bold("All data on the SD card", "will be lost")
+    return await hold_to_confirm(ctx, text, confirm="Format SD card")
+
+
 async def sd_problem_dialog(ctx: wire.GenericContext) -> bool:
     text = Text("SD card protection", ui.ICON_WRONG, ui.RED)
     text.normal("There was a problem", "accessing the SD card.")
     return await confirm(ctx, text, confirm="Retry", cancel="Abort")
 
 
-async def ensure_sd_card(ctx: wire.GenericContext) -> None:
-    sd = io.SDCard()
-    while not sd.present():
+async def ensure_sdcard(ctx: wire.GenericContext) -> None:
+    while not sdcard.is_present():
         if not await insert_card_dialog(ctx):
-            raise SdProtectCancelled("SD card required.")
+            raise SdCardUnavailable("SD card required.")
+
+    while True:
```

Comment:
this needs more work o_O
originally the `while True` was supposed to allow retry, but I don't think that is what happens currently